load-tests/src/load-test.ts

```
@@ -0,0 +1,284 @@
+    model_id: id,
+    name: 'test_plan' + rd + id,
+    start_time: plan_start_timestamp,
+    duration: "1d"
+  };
+
+  const resp = req.hasura(gql.CREATE_PLAN, {plan: input}, token, username);
+
+  sleep(Math.random() * 5);
+}
+
+export function insert_random_activities() {
+  const query = `#graphql
+    query {
+      plan {
+        id
+      }
+    }
+  `;
+
+  const rd = Math.random() * 10000;
+
+  const data = req.hasura(query, {}, token, username);
+  const { plan } = data;
+  const ids: number[] = plan;
+  const random_id = ids[Math.floor(Math.random() * ids.length)];
+
+  if (!random_id) {
+    sleep(5);
+    return;
+  }
+
+  const random_hour_offset = Math.floor(Math.random() * 24);
+
+  const input: ActivityInsertInput = {
+      arguments: {
+        biteSize: 1,
+      },
+      plan_id: id,
+      type: 'BiteBanana',
+      start_offset: random_hour_offset + 'h',
+  };
+
+  const resp = req.hasura(gql.CREATE_ACTIVITY_DIRECTIVE, {activityDirectiveInsertInput: input}, token, username);
+
+  sleep(Math.random() * 5);
+}
+
+export function insert_targeted_activities() {
+
+  const random_seconds_offset = Math.floor(Math.random() * 24 * 60 * 60);
+
+  const input: ActivityInsertInput = {
+      arguments: {
+        biteSize: 1,
+      },
+      plan_id: 1,
+      type: 'BiteBanana',
+      start_offset: random_seconds_offset + 's',
+  };
+
+  const resp = req.hasura(gql.CREATE_ACTIVITY_DIRECTIVE, {activityDirectiveInsertInput: input}, token, username);
+
+  sleep(Math.random() * 5);
+}
+
+export function run_targeted_simulations() {
+  const resp = req.hasura(gql.SIMULATE, {plan_id: 1}, token, username);
+  sleep(Math.random() * 30);
+}
+
+export function run_random_simulations() {
+  const query = `#graphql
+    query {
+      plan {
+        id
+      }
+    }
+  `;
+
+  const data = req.hasura(query, {}, token, username);
+  const { plan } = data;
+  const ids: number[] = plan;
+  const random_id = ids[Math.floor(Math.random() * ids.length)];
```

This line seems dangerous in the case that `plan` is empty (ie: there are no plans), as the code will try to get index 0 of `ids`, which is empty.

---

In that case, `random_id` will be `undefined` which we check for in the next line.

---

Can you alter the if statement so that if `ids` is empty, the method returns prior to setting `random_id`? That is actually what's being checked with `if(random_id)`

---

I've refactored this logic a bit, and I think I address your issue now.

---

I don't feel like my issue has been addressed, as my concern was how selecting the 0th element of an empty array meaning that we get an `undefined` wasn't clear, especially to primarily Java developers, where the expectation is an exception. In fact, I feel it may be worse, as the refactor now selects a property of an `undefined` object, which actually _will_ throw an exception. Additionally, `number` doesn't have properties, so there are some typing errors going on in these lines:

```ts
const data = resp.json("data");
const plan_ids: number[] = data.plan;
const random_id = plan_ids[Math.floor(Math.random() * plan_ids.length)].id;
```

An example of the errors I'm referring to can be found by attempting to run this snippet in the [Typescript Playground](https://www.typescriptlang.org/play), which is a simplification of the above three lines. 
```ts
const emptyList: number[] = []
const vary = emptyList[0].id;
console.log(vary)
```

---

I've added checks to catch several different errors now, which will show up in our statistics, rather than placing several `if === undefined` in the test functions themselves. I.e. we _want_ to fail loud in these scenarios, not gracefully handle errors.

How is the `number[]` type cast an error?

---

> I've added checks to catch several different errors now, which will show up in our statistics, rather than placing several if === undefined in the test functions themselves

I believe when I wrote this comment there were `!== undefined` checks in the test functions. The current error catching is good 👍 

> How is the `number[]` type cast an error?

This bit:

```ts
const random_plan = plan_ids[...];
return random_plan.id;
```

`plan_ids` is defined as `number[]` (which isn't the type of `data.plan`), which means that `random_plan` is of type `number`. The type `number` has no properties. Therefore, the fact that the code accesses the `.id` property of`random_plan` means that it isn't actually a `number`.

---

Ah duh... been starting at this code too long.